compiler/sem/liftdestructors.nim

```diff
@@ -1026,11 +1026,21 @@ proc createTypeBoundOps(g: ModuleGraph; c: PContext; orig: PType; info: TLineInf
   let skipped = orig.skipTypes({tyGenericInst, tyAlias, tySink})
   if isEmptyContainer(skipped) or skipped.kind == tyStatic: return
 
-  let h = sighashes.hashType(skipped, {CoType, CoDistinct})
-  var canon = g.canonTypes.getOrDefault(h)
-  if canon == nil:
-    g.canonTypes[h] = skipped
+  var canon: PType
+  if skipped.kind == tyObject:
+    # for nominal types, the type itself is already the canonical one (each one
+    # is unique)
+    # XXX: ^^ at present, this is only true for object types. Phantom
+    #      ``tyDistinct`` and ``tyEnum`` types still don't have unique
+    #      instances
     canon = skipped
+  else:
+    # structural types use canonicalization
+    let h = sighashes.hashType(skipped, {CoType, CoDistinct})
+    canon = g.canonTypes.getOrDefault(h)
+    if canon == nil:
+      g.canonTypes[h] = skipped
+      canon = skipped
```

Comment:
```suggestion
    # structural types use canonicalization
    canon = g.canonTypes.mgetOrPut(sighashes.hashType(skipped, {CoType, CoDistinct}), skipped)
```
(assuming that `nil` never gets put into `canonTypes`).

Reply:
Oh, very nice catch, thank you!